File: backend/src/main/java/com/cosplayinventory/backend/entity/InventoryItem.java
package com.cosplayinventory.backend.entity;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

import org.hibernate.annotations.UuidGenerator;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "inventory_items")
@Getter
@Setter
@NoArgsConstructor
public class InventoryItem {

    @Id
    @UuidGenerator
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "team_id", nullable = false)
    private Team team;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    private Product product;

    @Column(nullable = false)
    private BigDecimal quantity;

    @Column(nullable = false)
    private String unit;

    private BigDecimal price;

    @Column(name = "location_text")
    private String locationText;

    @Column(name = "expiry_date")
    private LocalDate expiryDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InventoryItemStatus status = InventoryItemStatus.SEALED;

    @Column(name = "opened_at")
    private LocalDate openedAt;

    @Column(name = "remaining_quantity")
    private BigDecimal remainingQuantity;

    @Column(name = "last_notified_date")
    private LocalDate lastNotifiedDate;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by", nullable = false)
    private User createdBy;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    void onCreate() {
        Instant now = Instant.now();
        this.createdAt = now;
        this.updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        this.updatedAt = Instant.now();
    }
}

File: app_mobile/lib/core/models/inventory_item.dart
import 'product.dart';

enum InventoryItemStatus { sealed_, opened, consumed, discarded }

InventoryItemStatus statusFromJson(String value) {
  switch (value) {
    case 'OPENED':
      return InventoryItemStatus.opened;
    case 'CONSUMED':
      return InventoryItemStatus.consumed;
    case 'DISCARDED':
      return InventoryItemStatus.discarded;
    default:
      return InventoryItemStatus.sealed_;
  }
}

String statusToJson(InventoryItemStatus status) {
  switch (status) {
    case InventoryItemStatus.opened:
      return 'OPENED';
    case InventoryItemStatus.consumed:
      return 'CONSUMED';
    case InventoryItemStatus.discarded:
      return 'DISCARDED';
    case InventoryItemStatus.sealed_:
      return 'SEALED';
  }
}

class InventoryItem {
  final String id;
  final Product product;
  final double quantity;
  final String unit;
  final double? price;
  final String? locationText;
  final DateTime? expiryDate;
  final InventoryItemStatus status;
  final DateTime? openedAt;
  final double? remainingQuantity;

  InventoryItem({
    required this.id,
    required this.product,
    required this.quantity,
    required this.unit,
    this.price,
    this.locationText,
    this.expiryDate,
    required this.status,
    this.openedAt,
    this.remainingQuantity,
  });

  factory InventoryItem.fromJson(Map<String, dynamic> json) => InventoryItem(
        id: json['id'] as String,
        product: Product.fromJson(json['product'] as Map<String, dynamic>),
        quantity: (json['quantity'] as num).toDouble(),
        unit: json['unit'] as String,
        price: (json['price'] as num?)?.toDouble(),
        locationText: json['locationText'] as String?,
        expiryDate: json['expiryDate'] != null ? DateTime.parse(json['expiryDate'] as String) : null,
        status: statusFromJson(json['status'] as String),
        openedAt: json['openedAt'] != null ? DateTime.parse(json['openedAt'] as String) : null,
        remainingQuantity: (json['remainingQuantity'] as num?)?.toDouble(),
      );
}
